src/bundle.c

```diff
@@ -171,30 +175,236 @@ static int unload_tracked_bundle(const char *bundle_name, struct list **subs)
+				required_by(reqd_by, bundle->component, mom, rl);
 			}
 		}
 	}
+}
+/* Return recursive list of included bundles */
+int show_included_bundles(char *bundle_name)
+{
+	int ret = 0;
+	int current_version = CURRENT_OS_VERSION;
+	struct list *subs = NULL;
+	struct list *deps = NULL;
+	struct manifest *mom = NULL;
+
+	current_version = get_current_version(path_prefix);
+	if (current_version < 0) {
+		fprintf(stderr, "Error: Unable to determine current OS version\n");
+		ret = ECURRENT_VERSION;
+		goto out;
+	}
+
+	mom = load_mom(current_version, false);
+	if (!mom) {
+		fprintf(stderr, "Cannot load official manifest MoM for version %i\n", current_version);
+		ret = EMOM_NOTFOUND;
+		goto out;
+	}
+
+	// add_subscriptions takes a list, so construct one with only bundle_name
+	struct list *bundles = NULL;
+	bundles = list_prepend_data(bundles, bundle_name);
+	ret = add_subscriptions(bundles, &subs, current_version, mom, true, 0);
+	list_free_list(bundles);
+	if (ret != add_sub_NEW) {
+		// something went wrong or there were no includes, print a message and exit
+		char *m = NULL;
+		if (ret & add_sub_ERR) {
+			string_or_die(&m, "Processing error");
+		} else if (ret & add_sub_BADNAME) {
+			string_or_die(&m, "Bad bundle name detected");
+		} else {
+			string_or_die(&m, "Unknown error");
+		}
+
+		fprintf(stderr, "Error: %s - Aborting\n", m);
+		free(m);
+		ret = 1;
+		goto out;
+	}
+	deps = recurse_manifest(mom, subs, NULL, false);
+	if (!deps) {
+		fprintf(stderr, "Error: Cannot load included bundles\n");
+		ret = ERECURSE_MANIFEST;
+		goto out;
+	}
+
+	/* deps now includes the bundle indicated by bundle_name
+	 * if deps only has one bundle in it, no included packages were found */
+	if (list_len(deps) == 1) {
+		fprintf(stderr, "No included bundles\n");
+		ret = 0;
+		goto out;
+	}
+
+	fprintf(stderr, "Bundles included by %s:\n\n", bundle_name);
+
+	struct list *iter;
+	iter = list_head(deps);
+	while (iter) {
+		struct manifest *included_bundle = iter->data;
+		iter = iter->next;
+		// deps includes the bundle_name bundle, skip it
+		if (strcmp(bundle_name, included_bundle->component) == 0) {
+			continue;
+		}
+
+		printf("%s\n", included_bundle->component);
+	}
+
+	ret = 0;
+
+out:
+	if (mom) {
+		free_manifest(mom);
+	}
+
+	if (deps) {
+		list_free_list_and_data(deps, free_manifest_data);
+	}
+
+	if (subs) {
+		free_subscriptions(&subs);
+	}
+
+	return ret;
+}
+
+int show_bundle_reqd_by(const char *bundle_name, bool server)
+{
+	int ret = 0;
+	int version = CURRENT_OS_VERSION;
+	struct manifest *current_manifest = NULL;
+	struct list *subs = NULL;
+	struct list *reqd_by = NULL;
+
+	if (!server && !is_tracked_bundle(bundle_name)) {
+		fprintf(stderr, "Error: Bundle \"%s\" does not seem to be installed\n", bundle_name);
+		fprintf(stderr, "       try passing --all to check uninstalled bundles\n");
+		ret = EBUNDLE_NOT_TRACKED;
+		goto out;
+	}
+
+	version = server ? get_latest_version() : get_current_version(path_prefix);
```

Comment:
My thought was that when `--all` is specified, we should show the dependency tree for the latest available version since they are asking for server dependencies. The use case for this is running bundle tests against the all dependent bundles in the latest version.

You are right though, for the more general case it makes more sense to report dependencies for the user's current version. The use case above is not that important since the testing will happen on the latest/staged version anyways.

Reply:
Thanks for clarifying. Yes, I think the bundle commands should only care about the current version, at least for now. In newer builds, there may be several new bundles available, and the user won't know about those bundles until they have updated anyway.